include/albatross/src/graph/minimum_spanning_tree.hpp

```diff
@@ -0,0 +1,174 @@
+std::set<VertexType>
+compute_vertices(const std::vector<Edge<VertexType>> &edges) {
+  std::set<VertexType> vertices;
+  for (const auto &edge : edges) {
+    vertices.insert(edge.a);
+    vertices.insert(edge.b);
+  }
+  return vertices;
+}
+
+template <typename VertexType> struct Graph {
+  std::vector<Edge<VertexType>> edges;
+  std::set<VertexType> vertices;
+
+  friend std::ostream &operator<<(std::ostream &os,
+                                  const Graph<VertexType> &graph) {
+    for (const auto &edge : graph.edges) {
+      std::cout << edge << std::endl;
+    }
+    return os;
+  }
+};
+
+template <typename VertexType>
+Graph<VertexType> create_graph(const std::vector<Edge<VertexType>> &edges) {
+  Graph<VertexType> graph;
+  graph.edges = edges;
+  graph.vertices = compute_vertices(edges);
+  return graph;
+}
+
+template <typename VertexType>
+void add_edge(const Edge<VertexType> &edge, Graph<VertexType> *graph) {
+  graph->edges.push_back(edge);
+  graph->vertices.insert(edge.a);
+  graph->vertices.insert(edge.b);
+}
+
+template <typename VertexType>
+bool contains_vertex(const Graph<VertexType> &graph, const VertexType &vertex) {
+  return graph.vertices.find(vertex) != graph.vertices.end();
+}
+
+template <typename VertexType>
+std::map<VertexType, std::vector<Edge<VertexType>>>
+adjacency_map(const Graph<VertexType> &graph) {
+  std::map<VertexType, std::vector<Edge<VertexType>>> output;
+
+  auto construct_or_push = [&output](const VertexType &key,
+                                     const Edge<VertexType> &new_edge) {
+    auto value = output.find(key);
+    if (value == output.end()) {
+      output[key] = {new_edge};
+    } else {
+      value->second.push_back(new_edge);
+    }
+  };
+
+  for (const auto &edge : graph.edges) {
+    construct_or_push(edge.a, edge);
+    Edge<VertexType> flipped(edge);
+    std::swap(flipped.a, flipped.b);
+    construct_or_push(edge.b, flipped);
+  }
+  return output;
+}
+
+template <typename VertexType>
+Graph<VertexType> maximum_spanning_tree(const Graph<VertexType> &graph) {
+  // https://en.wikipedia.org/wiki/Prim%27s_algorithm
+  // https://algs4.cs.princeton.edu/43mst/
+  // https://www.algotree.org/algorithms/minimum_spanning_tree/prims_c++/
+  Graph<VertexType> output;
+
+  const auto adjacency = adjacency_map(graph);
+
+  std::priority_queue<Edge<VertexType>> queue;
+
+  if (graph.edges.size() == 0) {
+    return output;
+  }
+
+  const Edge<VertexType> first_edge =
+      *std::max_element(graph.edges.begin(), graph.edges.end());
+
+  for (const auto &edge : adjacency.at(first_edge.a)) {
+    if (!contains_vertex(output, edge.b)) {
+      queue.push(edge);
+    }
+  }
+
+  double total_cost = 0;
+  while (!queue.empty()) {
+    const Edge<VertexType> next_edge = queue.top();
+    queue.pop();
+
+    const VertexType &next_vertex = next_edge.b;
+
+    if (!contains_vertex(output, next_vertex)) {
+      total_cost += next_edge.cost;
+      add_edge(next_edge, &output);
```

Comment:
I must be missing something - I don't see how this approach maximizes the span cost (unless maybe the edges are implicitly ordered by decreasing cost).

Not necessarily questioning the implementation, however, just wouldn't mind if you could point out what I might not be seeing.

Reply:
As a loose bit of related feedback, I find algorithmic/mathematical implementations in C++ _do_ benefit from regular use of comments to explain steps in the procedure. If this isn't the convention you follow in Albatross, then no harm done leaving it as-is.

Reply:
Yep, you got it.  The edges are placed in the `priority_queue` which orders them, so `const Edge<VertexType> next_edge = queue.top();` is the next largest edge.

In terms of comments I think I agree that this could benefit from some high level description of what is happening.  Mostly I try to stick to Swift's code style ... but there are other examples of complex parts of albatross which have comments.